src/Pirouette/Term/Transformations.hs

```diff
@@ -351,3 +358,176 @@ chooseHeadCase t ty args fstArg =
+    isTyVarInstance _ _ = Nothing
+
+    isArgInstance :: Int -> Int -> PrtArg -> PrtArg -> Maybe (M.Map String PrtArg)
+    isArgInstance nTe nTy (R.Arg tL) (R.Arg tT) = isInstance nTe nTy tL tT
+    isArgInstance nTe nTy (R.TyArg tyL) (R.TyArg tyT) = isTyInstance nTy tyL tyT
+    isArgInstance nTe nTy _ _ = Nothing
+
+    instantiate :: (MonadPirouette m)
+                => Int -> Int -> M.Map String PrtArg -> PrtTerm -> m PrtTerm
+    instantiate nTe nTy m tt@(R.App v@(R.F (FreeName x)) args) =do
+      case isHole x of
+        Nothing -> R.App v <$> mapM (instantiateArg nTe nTy m) args
+        Just i ->
+          case M.lookup i m of
+            Nothing -> throwError' $ PEOther $ "Variable " ++ show i ++ "_ appears on the right hand side, but not on the left-hand side."
+            Just (R.Arg t) ->
+              R.appN (shift (toInteger nTe) t) <$> mapM (instantiateArg nTe nTy m) args
+            Just (R.TyArg ty) -> throwError' $ PEOther $ "Variable x" ++ show i ++ "_ is at a term position on the right hand side, but was a type on the left-hand side."
+    instantiate nTe nTy m (R.App v args) =
+      R.App v <$> mapM (instantiateArg nTe nTy m) args
+    instantiate nTe nTy m (R.Lam ann ty t) =
+      R.Lam ann <$> instantiateTy nTy m ty <*> instantiate (nTe + 1) nTy m t
+    instantiate nTe nTy m (R.Abs ann k t) =
+      R.Abs ann k <$> instantiate nTe (nTy + 1) m t
+
+    instantiateTy :: (MonadPirouette m)
+                  => Int -> M.Map String PrtArg -> PrtType -> m PrtType
+    instantiateTy nTy m (R.TyApp v@(R.F (TyFree x)) args) =
+      case isHole x of
+        Nothing -> R.TyApp v <$> mapM (instantiateTy nTy m) args
+        Just i ->
+          case M.lookup i m of
+            Nothing -> throwError' $ PEOther $ "Variable " ++ show i ++ "_ appears on the right hand side, but not on the left-hand side."
+            Just (R.TyArg t) ->
+              R.appN (shift (toInteger nTy) t) <$> mapM (instantiateTy nTy m) args
+            Just (R.Arg ty) -> throwError' $ PEOther $ "Variable " ++ show i ++ "_ is at a type position on the right hand side, but was a term on the left-hand side."
+    instantiateTy nTy m (R.TyApp v args) =
+      R.TyApp v <$> mapM (instantiateTy nTy m) args
+    instantiateTy nTy m (R.TyLam ann k ty) =
+      R.TyLam ann k <$> instantiateTy (nTy + 1) m ty
+    instantiateTy nTy m (R.TyAll ann k ty) =
+      R.TyAll ann k <$> instantiateTy (nTy + 1) m ty
+    instantiateTy nTy m (R.TyFun a b) =
+      R.TyFun <$> instantiateTy nTy m a <*> instantiateTy nTy m b
+
+    instantiateArg :: (MonadPirouette m)
+                   => Int -> Int -> M.Map String PrtArg -> PrtArg -> m PrtArg
+    instantiateArg nTe nTy m (R.Arg t) = R.Arg <$> instantiate nTe nTy m t
+    instantiateArg nTe nTy m (R.TyArg ty) = R.TyArg <$> instantiateTy nTy m ty
+
+    isHole :: Name -> Maybe String
+    isHole x = head <$> matchRegex (mkRegex "([a-zA-Z][0-9]+)_") (T.unpack $ nameString x)
```

Comment:
Do we need to bring in an extra dependency for that? Can't we use words starting with an underscore and check whether length x > 1 && `head x == '_'`. Or, if you rather have underscores at the end, `length x > 1 && last x == '_'`

Reply:
Identifiers in PIR have to start by a letter, so if we keep using the underscore, we have to put it as the last character (or any other position, but the first). I agree that it is a dependency added for a quite restricted use of regular expressions. I can modify this.